### ISSUE TITLE
Adjusted Mutual Info metric, and what does iterations_per_epoch mean?

### ISSUE THREAD

**Issue by wconnell** [KevinMusgrave/pytorch-metric-learning#39]
I propose adding adjusted mutual information as a primary scoring metric. I recognize you highlight the many class NMI score inflation issue in "A Metric Learning Reality Check".  [scikit learn already has an implementation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.adjusted_mutual_info_score.html#sklearn.metrics.adjusted_mutual_info_score) so quite easy to add. Happy to discuss and help.

**Comment by KevinMusgrave** [comment]
I didn't know this was a thing! This is great, I'll let you know when I've added it. Also I'm thinking of converting the accuracies calculations to class form, so that you can override and add your own accuracy metrics.

**Comment by wconnell** [comment]
> Also I'm thinking of converting the accuracies calculations to class form

That's a good idea.


This can probably be moved to another issue but:
I'm having a bit of difficulty understanding the overall workflow. I think an overview would be greatly beneficial to the documentation. Maybe a jupyter notebook where losses/accuracies/embeddings are visualized. I'm finding no reduction in loss when I usually have good convergence using my own implementation. 

- Can you clarify what the `epochs` parameter means as compared to the `iterations_per_epoch`?? 
- How to use an embedding network without a trunk?
- Don't know how to interpret the tester accuracy logs are.

In a first pass the library looks really great - however I'm finding it difficult to troubleshoot under a very simple training scheme.

**Comment by KevinMusgrave** [comment]
Here's an [example training script](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/examples/example_MetricLossOnly.py) which might help. It comes with really useful logging if you ```pip install record-keeper tensorboard```. I could maybe convert these to jupyter notebooks, though I basically never use jupyter so I might need help with that.

1. Good question. I agree this can be confusing, and I actually just pushed a commit a few minutes ago which makes ```iterations_per_epoch``` optional. Here's how it works in the latest commit:

    - If you don't specify ```iterations_per_epoch``` then 1 epoch = 1 pass through the dataloader iterator. If ```sampler=None```, then 1 pass through the iterator is 1 pass through the dataset. If you use a sampler, then 1 pass through the iterator is 1 pass through the iterable returned by the sampler. This is basically fine for the standard samplers provided by PyTorch.
    - For samplers like MPerClassSampler or some offline mining method, the iterable returned might be very long or very short etc, and might not be related to the length of the dataset. The length of the epoch might vary each time the sampler creates a new iterable. In these cases, it can be useful to specify ```iterations_per_epoch``` so that each "epoch" is just a fixed number of iterations. The definition of epoch matters because there's various things like LR schedulers and hooks that depend on an epoch ending.

Here's the [latest code](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/trainers/base_trainer.py#L76) so you can see what I mean (I haven't pushed it to pypi or conda yet):
```python
        for self.epoch in range(start_epoch, num_epochs+1):
            self.set_to_train()
            logging.info("TRAINING EPOCH %d" % self.epoch)
            pbar = tqdm.tqdm(range(self.iterations_per_epoch))
            for self.iteration in pbar:
                self.forward_and_backward()
                self.end_of_iteration_hook(self)
                pbar.set_description("total_loss=%.5f" % self.losses["total_loss"])
                self.step_lr_schedulers(end_of_epoch=False)
            self.step_lr_schedulers(end_of_epoch=True)
            if self.end_of_epoch_hook(self) is False:
                break
```
Here's how ```iterations_per_epoch``` [is defined](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/trainers/base_trainer.py#L100) if you don't specify it:
```python
        if not self.iterations_per_epoch:
            self.iterations_per_epoch = len(self.dataloader)
```


2. A trunk is required. However, you can leave out the embedder from the ```models``` dictionary, and it'll work. And you don't have to pass in an embedder to the testers. The only purpose of "trunk" and "embedder" are to allow you to separate your model. But if you don't want to separate your model into 2 pieces, just pass in "trunk". As long as it outputs embedding vectors it'll work.


3. The epoch column refers to the accuracy of the model at a particular epoch during training. It's just to keep track of accuracy over time. Hopefully my answer above makes it clearer what epoch means.

Re: no reduction in loss, this can happen with some losses because they average their loss over only the non-zero elements, i.e. TripletMarginLoss and ContrastiveLoss do this by default. You can turn this off (**this will degrade performance**) by setting their init argument ```avg_non_zero_only``` to False.

If you're not seeing a reduction in loss, you can log the loss function properties at each iteration:

- num_non_zero_triplets for TripletMarginLoss
- num_non_zero_pos_pairs and num_non_zero_neg_pairs for ContrastiveLoss

If you use record-keeper and tensorboard, and the utils.HookContainer (as shown in the example scripts), these things will be logged automatically.

**Comment by KevinMusgrave** [comment]
Re: visualizing the embeddings, I don't have a great solution for that right now, but you can try this:

1. The BaseTester class has an init argument ```size_of_tsne```. Set this to some number less than or equal to the size of the dataset you're trying to visualize, and it'll compute tsne embeddings for a random subset of your dataset of size ```size_of_tsne```. You should probably set this to something low like 1000 or 2000, because right now it uses the sklearn implementation which is very slow. (I should have an option to just plot the whole dataset, but it's low priority right now.) I might also switch to UMAP since that is apparently faster.

2. You can access the tsne embeddings like this:
```python
tsne_embeddings = your_tester.tsne_embeddings
for split, sub_dict in tsne_embeddings.items():
    embeddings, labels = sub_dict[0]  # This extra sub dictionary is to allow tsne plots of datasets with multiple labels per dataset sample.
```
Then you can use something like matplotlib to plot. Alternatively, if you use utils.HookContainer, the [end of testing hook](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/utils/logging_presets.py#L65) will take care of this for you and save the tsne plot to tensorboard

**Comment by KevinMusgrave** [comment]
@wconnell v0.9.83.dev1 now has adjusted mutual info (AMI)., and accuracies are now computed using [utils.AccuracyCalculator](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/utils/accuracy_calculator.py#L46) which can be extended. 

For example, let's say you want to use the existing metrics but also compute precision @ 2, and a fancy mutual info method. You can extend the existing class, and write methods that start with the keyword ```calculate_```

```python
from pytorch_metric_learning.utils import accuracy_calculator, AccuracyCalculator

class YourCalculator(AccuracyCalculator):
    def calculate_precision_at_2(self, knn_labels, query_labels, **kwargs):
        return accuracy_calculator.precision_at_k(knn_labels, query_labels[:, None], 2)

    def calculate_fancy_mutual_info(self, query_labels, cluster_labels, **kwargs):
        return fancy_computations

    def requires_clustering(self):
        return super().requires_clustering() + ["fancy_mutual_info"] 
```

Any method that starts with "calculate_" will be passed the following kwargs:
```python
kwargs = {"query": query,                    # query embeddings
    "reference": reference,                     # reference embeddings
    "query_labels": query_labels,        
    "reference_labels": reference_labels,
    "embeddings_come_from_same_source": e,
    "label_counts": label_counts,           # a dictionary mapping from reference labels to the number of times they occur
    "knn_labels": knn_labels}                 # A 2d array where each row is the labels of the nearest neighbors of each query. The neighbors are retrieved from the reference set
```

If your method requires cluster labels, then append your method's name to the ```requires_clustering``` list, via ```super()```. Then, if any of your methods need cluster labels, ```self.get_cluster_labels()``` will be called, and the kwargs will include ```cluster_labels```.

Now when ```get_accuracy``` is called, the returned dictionary will contain ```precision_at_2``` and ```fancy_mutual_info```:
```python
calculator = YourCalculator()
acc_dict = calculator.get_accuracy(query_embeddings,
    reference_embeddings,
    query_labels,
    reference_labels,
    embeddings_come_from_same_source=True
)
# Now acc_dict contains the key "precision_at_2" and "fancy_mutual_info".

```

You can use your custom calculator with the tester classes as well, by passing it in as an init argument:
```python
from pytorch_metric_learning import testers
t = testers.GlobalEmbeddingSpaceTester(..., accuracy_calculator=YourCalculator())
```

The existing calculator computes both NMI and AMI. Let's say you don't want to bother computing NMI. You can exclude metrics, either when initializing the calculator, and/or when calling ```get_accuracy```:
```python
calculator = YourCalculator(exclude_metrics=("NMI")) # when you init
calculator.get_accuracy(...)

# and/or when you call get_accuracy
calculator = YourCalculator()
calculator.get_accuracy(..., exclude_metrics=("NMI"))
```
